Use ocr-highlighting plugin with MiniOCR

albig opened the pull request:
Just a PR to discuss the current state.

beatrycze-volk committed:
Introduce solr highlighting

beatrycze-volk committed:
Use config for index remapped in SearchInDocumentTool

beatrycze-volk committed:
Use highlighting feature

beatrycze-volk committed:
Update js file to use new OCR highlighting

beatrycze-volk committed:
Change submit in search to click

beatrycze-volk committed:
Move search to own method

beatrycze-volk committed:
Update methods in JS of search in document plugin

beatrycze-volk committed:
Change configuration option fromisIndexRemapped to searchURL

Actually isIndexRemapped is not needed as fields are using config variables which have defined default values equal to fields names in default SOLR schema of Kitodo Presentation.

beatrycze-volk committed:
Make search input names configurable

beatrycze-volk committed:
Use config for documentIdUrlSchema

It is needed for cases when document id is passed by url instead by simple id used in index or database.

beatrycze-volk committed:
Update JS ad PHP files to use configured input names and document id URL schema, implement correctly
displaying of the links

beatrycze-volk committed:
Fix errors pointed by Codacy

beatrycze-volk committed:
Add max amount of snippets and restore query->getHighlighting

Plugin has return only one snippet per page as default. Restore getHighlighting without assigning it to the variable.

beatrycze-volk committed:
Adjust URL generation for results' links

After this adjustments displaying and highlighting of the results in image works.

beatrycze-volk committed:
Add event listener for Enter pressed in the search field

beatrycze-volk committed:
Restore usage of submit button for submitting search

beatrycze-volk committed:
Restore CORS header in PageViewProxy

beatrycze-volk committed:
Add missing search component to SOLR config

beatrycze-volk committed:
Check if passed id is number or numeric string

Previous check was converting numeric string to int if it was starting with number

beatrycze-volk committed:
Remove indexing of full text from file

Source pointer can't be on any other server han SOLR, so it is not usable here.

beatrycze-volk committed:
Save to index full OCR

+ add working logging for document and indexing

beatrycze-volk committed:
Remove getRawText methods

beatrycze-volk committed:
Return false from save method if documents were not indexed

Either because of incorrect solr core or because some errors while indexing

beatrycze-volk committed:
Fix after rebase class for logger in Document instance

beatrycze-volk committed:
Highlight more than one word in search

beatrycze-volk committed:
Preserve search phrase and get hit list after hit link was clicked

beatrycze-volk committed:
Index documents with empty fulltext field

beatrycze-volk committed:
Use coordinates inside the function for searching full text features

beatrycze-volk committed:
Fix Codacy errors

beatrycze-volk committed:
Fix request handler in solr config

beatrycze-volk committed:
Use new syntax for getting of configuration

beatrycze-volk committed:
Fix documentations

beatrycze-volk committed:
Update Resources/Public/Javascript/PageView/PageView.js

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

beatrycze-volk committed:
Check for undefined baseUrl and remove of passing hl param for highligting

beatrycze-volk committed:
Add checks for empty query parameters

Push them only if they are actually existing.

kselmi committed:
Update README.md

added: information on communication and support structure; demo server

beatrycze-volk committed:
Fix rebase errors

Commit:
Merge branch 'master' into use-solr-highlighting

beatrycze-volk committed:
Update Classes/Plugin/Tools/SearchInDocumentTool.php

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

beatrycze-volk committed:
Reverse not needed changes

beatrycze-volk committed:
Add comment for documentIdUrlSchema

beatrycze-volk committed:
Remove reading of URL parameters for hl in image

Now image hl is made by coordinates, so passing word from URL parameter is not needed anymore.

beatrycze-volk committed:
Use SOLR plugin in the document list

Commit:
Merge branch 'master' into use-solr-highlighting

beatrycze-volk committed:
Fix errors and add missing docs

beatrycze-volk committed:
Merge remote-tracking branch 'origin/use-solr-highlighting' into use-solr-highlighting

beatrycze-volk committed:
Fix Scrutinizer warnings

beatrycze-volk committed:
Apply review comments related to URL parameters

beatrycze-volk committed:
Remove overwrite of query

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

beatrycze-volk committed:
Use create filter query

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

beatrycze-volk committed:
Fix the query limit

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

Commit:
Update Classes/Common/Indexer.php

This line is not needed.

beatrycze-volk committed:
Ignore false-positive warning from Scrutinizer

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

beatrycze-volk committed:
Add facet component to solrconfig.xml

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

beatrycze-volk committed:
Use array for filter queyr

Co-authored-by: Alexander Bigga <alexander.bigga@slub-dresden.de>

Commit:
Merge branch 'master' into use-solr-highlighting

beatrycze-volk committed:
Fix HTML generation for full text

beatrycze-volk committed:
Remove Illustration tag from full text XML

Commit:
Merge branch 'master' into use-solr-highlighting

Commit:
Merge pull request #587 from beatrycze-volk/use-solr-highlighting

Use Solr OCR Highlighting Plugin in Search in Document Plugin

Commit:
Merge pull request #639 from beatrycze-volk/remove-obosolete-hl-parameter

Remove obsolete hl parameter use in image highlighting

Commit:
Revert "Remove Illustration tag from full text XML"

This reverts commit 66bbed3074bbd73e54e0c1469e9d5bfef3a21b39.

Commit:
Move the logger output to the right position

Commit:
It could happen, that $snippet is an empty array

Commit:
Keep the old code to support different text formats and classes.

Commit:
Use MiniOCR format to store fulltext to ocr-highlighting plugin.

This supports only ALTO 2.x.

There should be found a solution to support different namespaces. See https://github.com/kitodo/kitodo-presentation/issues/488

beatrycze-volk commented on Classes/Format/Alto.php:
```diff
@@ -42,4 +46,49 @@ public function getRawText(\SimpleXMLElement $xml)
+     */
+    public function getTextAsMiniOcr(\SimpleXMLElement $xml)
+    {
+        $rawText = '';
```
There is a complain from Scrutinizer about not used `$rawText` variable. It needs to to be removed or used in return in case blocks are empty.

albig replied:
Thank you. That's fixed now. This was taken from the `getRawText()` method, I've modified before.

beatrycze-volk commented on Classes/Format/Alto.php:
```diff
@@ -42,4 +46,49 @@ public function getRawText(\SimpleXMLElement $xml)
+    public function getTextAsMiniOcr(\SimpleXMLElement $xml)
+    {
+        $rawText = '';
+        $xml->registerXPathNamespace('alto', 'http://www.loc.gov/standards/alto/ns-v2#');
```
There are already documents with version 3 and 4. If I'm not mistaken Zeitungsportal has mostly them. How are we going to handle those documents?

albig replied:
Of course, there is ALTO v3 and v4. But this is a another issue (#488) which has to be fixed. Maybe by different Alto classes with different "getTextAsMiniOcr()" methods.

For the DDB-Zeitungsportal is no problem here as we do not index into their Solr.

beatrycze-volk replied:
Good point! Thanks! Then there is no problem.

beatrycze-volk commented on Classes/Format/Alto.php:
```diff
@@ -42,4 +46,49 @@ public function getRawText(\SimpleXMLElement $xml)
+                    foreach ($value->children() as $lkey => $word) {
+                        if ($lkey == "String") {
+                            $attributes = $word->attributes();
+                            $la =  Solr::escapeQuery((string) $attributes['CONTENT']);
```
What is the purpose of this variable?

albig replied:
Nothing. This was a remaining debugging helper. I removed it.

beatrycze-volk commented on Classes/Format/Alto.php:
```diff
@@ -42,4 +46,49 @@ public function getRawText(\SimpleXMLElement $xml)
+            }
+        }
+
+        return $miniOcr->asXml();
```
Shouldn't we note that in the case of error of the document creation this method can return false instead of string?

albig replied:
Not sure, what the ocr highlighting plugin will accept. I think, we must send an empty string to it. 

Could be improved for sure.

albig replied:
But you're right. We shouldn't return the output of `asXml()` directly.

beatrycze-volk commented on Classes/Format/Alto.php:
```diff
@@ -42,4 +46,49 @@ public function getRawText(\SimpleXMLElement $xml)
         }
         return $rawText;
     }
+
```
Probably only one empty line is needed here.

Commit:
Fix scrutinizer hints

beatrycze-volk committed:
Fix check for uid and add missing doc comments

Commit:
Merge branch 'master' into ocr-highlighting-plugin

Commit:
Add sorting which has been lost by solving merge conflicts

beatrycze-volk committed:
Ad missing docs and small improvments

beatrycze-volk committed:
Merge remote-tracking branch 'kitodo/ocr-highlighting-plugin' into ocr-highlighting-plugin

albig commented on Classes/Format/Alto.php:
```diff
@@ -42,4 +46,75 @@ public function getRawText(\SimpleXMLElement $xml)
+        }
+
+        $miniOcrXml = $miniOcr->asXml();
+        if(\is_string($miniOcrXml)) {
```
```suggestion
        if (\is_string($miniOcrXml)) {
```

Commit:
Update Classes/Format/Alto.php

Commit:
Fix scrutinizer hint

Commit:
Fix scrutinizer hint

scrutinizer-auto-fixer committed:
Scrutinizer Auto-Fixes

This commit consists of patches automatically generated for this project on https://scrutinizer-ci.com

beatrycze-volk committed:
Fix defintion of textMiniOcr variable

beatrycze-volk committed:
Remove double escape of special characters

beatrycze-volk committed:
Remove unused getRawText method

beatrycze-volk committed:
Extract reading of the xml object to own method

beatrycze-volk committed:
Change return of add method in Indexer

It is counter intuitve to return 0 on success and 1 on failure. It can cause mistakes when return of this method is actually in use.

beatrycze-volk reviewed (dismissed):
I have looked over the code and tested changes. It seems to be cleaned out and working properly (indexing of the documents and searching in the document).

Commit:
Merge branch 'ocr-highlighting-plugin' into scrutinizer-patch-1

Commit:
Merge pull request #674 from kitodo/scrutinizer-patch-1

Scrutinizer Auto-Fixes

Commit:
Revert "Remove unused getRawText method"

This reverts commit f0d764545c58a8de2a0d07b1c9814b585a97b189.

Commit:
Remove unused use statement

Commit:
Set recommended parameter if page nummber is already known.

Commit:
Fix sorting in listview

Commit:
Update installation instruction for the ocr highlighting plugin

Commit:
Merge branch 'master' into fix-id-in-fulltext

Commit:
Merge pull request #649 from beatrycze-volk/fix-id-in-fulltext

Fix HTML generation for full text

Commit:
Merge branch 'master' into patch-2

Commit:
Merge branch 'ocr-highlighting-plugin' into patch-2

Commit:
Merge pull request #637 from kselmi/patch-2

Update README.md

Commit:
Fix fulltext js and invalid textsnippet markup

Commit:
Merge branch 'ocr-highlighting-plugin' of github.com:kitodo/kitodo-presentation into ocr-highlighting-plugin

Commit:
Merge branch 'master' into ocr-highlighting-plugin

Commit:
There is no version 3.0 of Solr schema.

beatrycze-volk committed:
Highlight search result on the base of begin position

Ignore width and height as it can differ with results returned from OCR plugin

beatrycze-volk committed:
Simplify id creation for full text

Create id only from begin position of the text
